### ISSUE TITLE
hulu doesn't work

### ISSUE THREAD

**Issue by GoogleCodeExporter** [MiraculousLegendas/get-flash-videos#249]
```
What steps will reproduce the problem?
1. try to download a video from hulu
2. watch it fail

What is the expected output? What do you see instead?
Expected that it will download.  Here's what I see instead:

$ get_flash_videos-1.24 -d 
http://www.hulu.com/watch/212550/white-collar-countermeasures
1 plugin installed:
- Hulu.pm
Downloading http://www.hulu.com/watch/212550/white-collar-countermeasures
-> GET http://www.hulu.com/watch/212550/white-collar-countermeasures
<- 200 text/html; charset=utf-8 (174180): UTF8 on, non-ASCII, 174180 characters 
174184 bytes
Trying to open plugin ~/.get_flash_videos/plugins/Www.pm
Trying to open plugin ~/.get_flash_videos/plugins/Hulu.pm
Using method 'hulu' for 
http://www.hulu.com/watch/212550/white-collar-countermeasures
Found Hulu CID: 50119747
Returning cached Hulu data
-> GET http://r.hulu.com/videos?content_id=50119747
<- 200 application/xml; charset=utf-8 (6959): UTF8 on, ASCII, 6959 characters 
6959 bytes
Found Hulu encrypted PID: NO_MORE_RELEASES_PLEASE_50119747
Decrypted Hulu PID: 'NO_MORE_RELEASES_PLEASE_50119747'
Auth: 63532eb7aace4c8819a0e8fc7e523490
Hulu SMIL URL: 
http://s.hulu.com/select.ashx?pid=NO_MORE_RELEASES_PLEASE_50119747&auth=63532eb7
aace4c8819a0e8fc7e523490&v=713434170&np=1
-> GET 
http://s.hulu.com/select.ashx?pid=NO_MORE_RELEASES_PLEASE_50119747&auth=63532eb7
aace4c8819a0e8fc7e523490&v=713434170&np=1
<- 400 text/html (210): UTF8 off, ASCII, 210 characters 210 bytes
Encrypted XML: '<html>
<head>
<title>400 Bad Request</title>
</head>
<body>
    <h2>400 Bad Request</h2>
    <p>pid 'NO_MORE_RELEASES_PLEASE_50119747' does not map to asset id</p>
    <pre id="traceback"></pre>
</body>
</html>'
XML decrypt key: 
4878B22E76379B55C962B18DDBC188D82299F8F52E3E698D0FAF29A40ED64B21, IV: 
WA7hap7AGUkevuth
Error: encrypt: datasize not multiple of blocksize (16 bytes) at 
/loader/0x19a2100/FlashVideo/Site/Hulu.pm line 456.

Couldn't extract Flash movie URL. This site may need specific support adding,
or fixing.

Please confirm the site is using Flash video and if you have Flash available
check that the URL really works(!).

Check for updates by running: ./combined-get_flash_videos-1.24 --update

If the latest version does not support this please open a bug (or
contribute a patch!) at http://code.google.com/p/get-flash-videos/
make sure you include the output with --debug enabled.
Couldn't download any videos.

What version of the product are you using? On what operating system?

latest version, linux

Please provide any additional information below.

Is Hulu getting wise on this script and asking that "no more releases" are 
created??

Found Hulu encrypted PID: NO_MORE_RELEASES_PLEASE_50119747
```

Original issue reported on code.google.com by `imon...@hmamail.com` on 12 Mar 2011 at 5:59

**Comment by GoogleCodeExporter** [comment]
```
Any update regarding this?

I reported it here, as well:
http://code.google.com/p/get-flash-videos/issues/detail?id=135#c4
```

Original comment by `menachem...@gmail.com` on 15 Mar 2011 at 11:47

**Comment by GoogleCodeExporter** [comment]
```
I've had the same issue. Has anyone figured out of anyother way to get the 
"pid"?

It's not the number is it?
```

Original comment by `mjbauer95` on 16 Mar 2011 at 12:42

**Comment by GoogleCodeExporter** [comment]
```
If you've just been following this (like me), you may not know that the Hulu 
Plugin is officially broken at the moment. 

To quote the plugin page: 
http://gitorious.org/get-flash-videos-plugins/pages/Hulu
--------------
Plugin currently broken (March 2011)

The plugin is currently broken due to the various CDNs used by Hulu using a 
different type of Flash RTMPE handshake, which means rtmpdump can’t download 
Hulu’s streams. This also breaks other applications and downloads from other 
sites. We don’t know when this will be fixed. rtmpdump is a volunteer 
project. It also seems like Hulu has changed various aspects of how their site 
works, breaking this plugin even more.
--------------

In other words, there are two issues preventing this plugin from working at the 
moment. The first is the changes hulu have made to their backend (as well as 
this issue):
http://code.google.com/p/get-flash-videos/issues/detail?id=243


The second is the handshake 9 protocol, which rtmpdump does not yet support. 
This has affected many sites, not just hulu:
http://lists.mplayerhq.hu/pipermail/rtmpdump/2011-March/001304.html

However, I did come across this forum post, where the user claimed to be able 
to get sites that did not work with the latest version of rtmpdump (due to the 
handshake 9 issue) to work with an older version of rtmpdump, version 2.1d.
http://stream-recorder.com/forum/showpost.php?s=74eec2adf543f7b4ca65d65df2d32e47
&p=26332&postcount=5

Maybe this version of rtmpdump would work with hulu as well.
```

Original comment by `menachem...@gmail.com` on 22 Mar 2011 at 4:12

**Comment by GoogleCodeExporter** [comment]
```
Looking at the rtmpdump changelog for v2.1d > v2.2 I see "more FlashPlayer 10 
handshake support". So it looks like handshake testing added in later releases 
of rtmpdump is causing it to shoot itself in the foot.

I see that Howard Chu has submitted a SVN change for handshake work. So I guess 
the work is underway.

"Author: hyc
Date: Sun Mar 20 00:32:53 2011
New Revision: 562

Log:
Drop back on the handshake version, avoid rtmpe 9

Modified:
   trunk/librtmp/handshake.h"
```

Original comment by `pcwal...@comcast.net` on 22 Mar 2011 at 5:40

**Comment by GoogleCodeExporter** [comment]
```
Issue 259 has been merged into this issue.
```

Original comment by `mjbauer95` on 17 May 2011 at 2:17

**Comment by GoogleCodeExporter** [comment]
Original comment by `mjbauer95` on 17 May 2011 at 8:39
- Changed state: **Accepted**
- Added labels: **Priority-High**
- Removed labels: **Priority-Medium**

**Comment by GoogleCodeExporter** [comment]
```
rtmpdump supports rtmpe 9 handshakes now.
```

Original comment by `Highland...@gmail.com` on 13 Jul 2011 at 11:23

**Comment by GoogleCodeExporter** [comment]
```
Any update on Hulu?
```

Original comment by `KanK...@gmail.com` on 29 Jul 2011 at 9:01

**Comment by GoogleCodeExporter** [comment]
```
I've got a VERY rough patch based on the xmbc hulu plugin. It should be cleaned 
up.
This only works with new rtmpdump.

```

Original comment by `tmavroi...@gmail.com` on 30 Jul 2011 at 2:33

Attachments:
- [hulu.patch](https://storage.googleapis.com/google-code-attachments/get-flash-videos/issue-249/comment-9/hulu.patch)

**Comment by GoogleCodeExporter** [comment]
```
Fixed - updated code is on Gitorious. Running get_flash_videos --update should 
get you the latest version. The plugin's wiki page has been updated:
https://gitorious.org/get-flash-videos-plugins/pages/Hulu

Subtitles have also been fixed.

Many thanks to tmavroi for the patch and to hyc for updating rtmpdump.
```

Original comment by `zakflash...@gmail.com` on 3 Aug 2011 at 6:43
- Changed state: **Verified**

**Comment by GoogleCodeExporter** [comment]
```
I ran the update, upgraded to the latest version of rtmpdump, but still ran 
into a problem.

I am running:
Ubuntu 11.04

RTMPDump v2.4 (retrieved using 'git clone git://git.ffmpeg.org/rtmpdump' and 
installed using 'make SYS=posix' followed by 'sudo make install')

get_flash_video 1.24 (Ran get_flash_video --update to get latest hulu plugin as 
well)


Here is the command I ran:
--
get_flash_videos --debug 
http://rss.hulu.com/~r/HuluRecentlyAddedVideos/~3/sOb9ynAsHo4/attack-of-the-show
-felicia-day-eureka-interview 2>&1|tee log
--

I got an error message saying, see the attached log:
--
Download failed, no valid file downloaded
Couldn't download any videos.
--
```

Original comment by `menachem...@gmail.com` on 3 Aug 2011 at 7:15

Attachments:
- [log](https://storage.googleapis.com/google-code-attachments/get-flash-videos/issue-249/comment-11/log)

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
I get the same results as comment 11.

Ubuntu 11.4
RTMPDump 2.4
get_flash_videos version 1.24
updated to latest Hulu.pm, dated 2011-08-03

command run:
get_flash_videos --debug 
http://www.hulu.com/watch/64189/knight-rider-classic-short-notice 2>&1 | tee 
hulu_log

hulu_log attached.
```

Original comment by `jsha...@gmail.com` on 4 Aug 2011 at 8:56

Attachments:
- [hulu_log](https://storage.googleapis.com/google-code-attachments/get-flash-videos/issue-249/comment-13/hulu_log)

**Comment by GoogleCodeExporter** [comment]
```
I just ran 'get_flash_videos --update'. It didn't seem to run any updates, but 
now the videos from Comment 11 and 13 both worked.
```

Original comment by `menachem...@gmail.com` on 4 Aug 2011 at 10:53

**Comment by GoogleCodeExporter** [comment]
```
I have get_flash_videos version 1.24 and have the same problem as comment 11 
and 13.
```

Original comment by `pcwal...@comcast.net` on 5 Aug 2011 at 2:35

**Comment by GoogleCodeExporter** [comment]
```
Did you upgrade to the latest version of RTMPDump as well?
```

Original comment by `menachem...@gmail.com` on 5 Aug 2011 at 2:38

**Comment by GoogleCodeExporter** [comment]
```
Yes, I downloaded the latest git
```

Original comment by `pcwal...@comcast.net` on 5 Aug 2011 at 5:32

**Comment by GoogleCodeExporter** [comment]
```
Debug log with latest hulu plugin and rtmpdump 2.4


```

Original comment by `pcwal...@comcast.net` on 6 Aug 2011 at 2:03

Attachments:
- [log](https://storage.googleapis.com/google-code-attachments/get-flash-videos/issue-249/comment-18/log)

**Comment by GoogleCodeExporter** [comment]
```
re: #18 - your log clearly says rtmpdump v2.3 so you obviously did not install 
rtmpdump 2.4.
```

Original comment by `Highland...@gmail.com` on 6 Aug 2011 at 10:09

**Comment by GoogleCodeExporter** [comment]
```
Yes, my mistake. However, I uploaded an old log.

Since this issue has been closed, I opened a new issue #293 with the rtmpdump 
2.4 log error.


```

Original comment by `pcwal...@comcast.net` on 6 Aug 2011 at 8:17
